lib/improver/database/spotdata_database.py

```diff
@@ -0,0 +1,276 @@
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import iris
+import iris.iterate
+import iris.pandas
+import itertools
+import numpy as np
+import pandas as pd
+import sqlite3
+import os
+from datetime import datetime as dt
+from pandas import DataFrame
+
+
+# For debugging
+from profiler import profile
+import pprint
+import resource
+
+
+class SpotDatabase(object):
+
+    def __init__(self, cubelist,
+                 primary_dim='time',
+                 primary_map=['validity_date', 'validity_time'],
+                 primary_func=[lambda x:dt.utcfromtimestamp(x).date(),
+                               lambda x:dt.utcfromtimestamp(x).hour*100],
+
+                 pivot_dim='forecast_period',
+                 pivot_map=lambda x: 'fcr_tplus{:03d}'.format(int(x/3600)),
+
+                 column_dims=['wmo_site', 'name'],
+                 column_maps=['station_id', 'cf_name']):
+
+        """
+        Initialise class.
+
+        """
+
+        self.cubelist = cubelist
+        self.pivot_dim = pivot_dim
+        self.pivot_map = pivot_map
+        self.primary_dim = primary_dim
+        self.primary_map = primary_map
+        self.primary_func = primary_func
+
+        self.column_dims = column_dims
+        self.column_maps = column_maps
+        self.column_func = []
+
+        self.assert_similar()
+
+    def __repr__(self):
+        """
+        Representation of the instance.
+
+        """
+
+        return '<SpotDatabase: {}>'.format(self.primary_dim)
+
+    def assert_similar(self):
```

Comment:
This is a nice little function. It probably just needs two tests, one where they are comparable and one where they are not and it throws an exception.

Reply:
It does get slow when you have lots of cubes in your cubelist